cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/RoleControllerSpi.java

```diff
@@ -0,0 +1,76 @@
+package io.cdap.cdap.security.spi.authorization;
+
+import io.cdap.cdap.api.security.AccessException;
+import io.cdap.cdap.proto.security.GrantedPermission;
+import io.cdap.cdap.proto.security.Principal;
+import io.cdap.cdap.proto.security.Role;
+
+import java.util.Map;
+
+/**
+ * This interface is responsible to manage role related actions and it's relation with {@link Principal}
+ *
+ * This logic was previously used in {@link AccessController}
+ * This is an SPI .
+ */
+public interface RoleControllerSpi {
+
+  /**
+   * Create a role.
+   *
+   * @param role the {@link Role} to create
+   * @throws AlreadyExistsException if the the role to be created already exists
+   */
+  AuthorizationResult createRole(Role role) throws AccessException;
+
+  /**
+   * Drop a role.
+   *
+   * @param role the {@link Role} to drop
+   * @throws AlreadyExistsException if the role to be dropped is not found
+   */
+  AuthorizationResult dropRole(Role role) throws AccessException;
+
+  /**
+   * Add a role to the specified {@link Principal}.
+   *
+   * @param role the {@link Role} to add to the specified group
+   * @param principal the {@link Principal} to add the role to
+   * @throws AlreadyExistsException if the role to be added to the principals is not found
+   */
+  AuthorizationResult addRoleToPrincipal(Role role, Principal principal) throws AccessException;
+
+  /**
+   * Delete a role from the specified {@link Principal}.
+   *
+   * @param role the {@link Role} to remove from the specified group
+   * @param principal the {@link Principal} to remove the role from
+   * @throws AlreadyExistsException if the role to be removed to the principals is not found
+   */
+  AuthorizationResult removeRoleFromPrincipal(Role role, Principal principal) throws AccessException;
+
+  /**
+   * Returns a set of all {@link Role roles} for the specified {@link Principal}.
+   *
+   * @param principal the {@link Principal} to look up roles for
+   * @return Set of {@link Role} for the specified {@link Principal}
+   */
+  Map<Role, AuthorizationResult> listRoles(Principal principal) throws AccessException;
```

Comment:
Also here. You don't give each role an Authorization result, you have a single result for the call